feat(source): adding metric ingestion to opentelemetry source

cmcmacs opened the pull request:
## Summary
This PR adds ability to accept metrics via Opentelemetry Source.
Due to challenges in mapping OpentelMetrics to Internal Vector DataModel, there are different mechanics to create the mapping, but overall the short view:
OpenTelemetry Gauge => Vector Gauge(MetricKind::Absolute)
OpenTelemetry Sum => Vector Counter(if is_monotonic, then MetricKind::Incremental else MetricKind::Absolute)
OpenTelemetry Histogram => Vector AggregatedHistogram
OpenTelemetry Exponential Histogram => Vector AggregatedHistogram <- this is pretty strange, but I don't know how to do it better, we basically restore bucket ranges from scale
OpenTelemetry Summary(deprecated) => Vector Aggregated Summary

Doc for reference: https://opentelemetry.io/docs/specs/otel/metrics/data-model/

## Change Type
- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## How did you test this PR?
<!-- Please describe your testing plan here.
Sharing information about your setup and the Vector configuration(s) you used (when applicable) is highly recommended.
Providing this information upfront will facilitate a smoother review process. -->

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist
- [x] Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
  - `make check-all` is a good command to run locally. This check is
    defined [here](https://github.com/vectordotdev/vector/blob/1ef01aeeef592c21d32ba4d663e199f0608f615b/Makefile#L450-L454). Some of these
    checks might not be relevant to your PR. For Rust changes, at the very least you should run:
    - `cargo fmt --all`
    - `cargo clippy --workspace --all-targets -- -D warnings`
    - `cargo nextest run --workspace` (alternatively, you can run `cargo test --all`)
- [ ] If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `dd-rust-license-tool write` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

## References

- Closes: [https://github.com/vectordotdev/vector/issues/17309](https://github.com/vectordotdev/vector/issues/17309)

pront commented on changelog.d/opentelemetry_metrics.feature.md:
```diff
@@ -0,0 +1,3 @@
+Support ingesting metrics through opentelemetry source(experimental)
```
```suggestion
The `opentelemetry` source now supports metrics ingestion.
```

pront commented on website/cue/reference/components/sources/opentelemetry.cue:
```diff
@@ -41,7 +41,7 @@ components: sources: opentelemetry: {
 		requirements: []
 		warnings: [
```
We can now remove this warning completely :)

cmcmacs replied:
removed

pront commented on website/cue/reference/components/sources/opentelemetry.cue:
```diff
@@ -256,6 +262,12 @@ components: sources: opentelemetry: {
+		metrics: {
+			title: "Ingest metrics"
+			body: """
+				Metrics support is experimental and subject to change due to the differences in metrics structure between internal Vector metric DataModel and OpenTelemetry, difference to note: Exponential Histogram is transformed into regular Aggregated Histogram, buckets are being calculated based on the base and scale.
```
👍 I think we embellish this section with the mappings you provided in the PR description.

pront commented on lib/opentelemetry-proto/src/convert.rs:
```diff
@@ -57,6 +64,159 @@ impl ResourceLogs {
     }
 }
 
+impl ResourceMetrics {
```
Nit: We can separate logs and metrics conversion logic. Same for the `src/sources/opentelemetry/tests.rs`. You can leave this as is though, up to you. This suggestion is purely for better maintainability.

pront commented on lib/opentelemetry-proto/src/convert.rs:
```diff
@@ -57,6 +64,159 @@ impl ResourceLogs {
 
+impl ResourceMetrics {
+    pub fn into_event_iter(self) -> impl Iterator<Item = Event> {
+        let resource = self.resource.clone();
```
```suggestion
        let resource = self.resource.as_ref();
```

It should be possible to clone when we create the Vector event.

cmcmacs replied:
I tried this, and it's a bit hard to manage due to lifetime restrictions

cmcmacs replied:
we can prob switch to using Arc to save on cloning

cmcmacs commented:
I changed unwrap to unwrap_or for sum and gauges to prevent panics on None if something not-expected to happen.

cmcmacs commented:
> Hi @cmcmacs, this looks great. Thanks for adding the tests as well. One general question, did you test this end-to-end with an OTEL exporter publishing to Vector?

I did test it with metrics and traces. Used sample python to push, here is an example:
```
import time
import random
from opentelemetry import metrics
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter

resource = Resource(attributes={
    "service.name": "demo-otel-metrics-client",
    "service.instance.id": "instance-42",
    "host.name": "localhost",
})

exporter = OTLPMetricExporter(endpoint="localhost:4317", insecure=True)
reader = PeriodicExportingMetricReader(exporter, export_interval_millis=5000)

provider = MeterProvider(resource=resource, metric_readers=[reader])
metrics.set_meter_provider(provider)
meter = metrics.get_meter(__name__)

request_counter = meter.create_counter(
    name="example.counter",
    description="A monotonic counter",
    unit="1"
)

connections = meter.create_up_down_counter(
    name="example.updowncounter",
    description="An up-down counter for connections",
    unit="connections"
)

response_times = meter.create_histogram(
    name="example.histogram",
    description="Histogram of response times",
    unit="ms"
)

def observable_gauge_callback(options):
    return [metrics.Observation(value=random.uniform(0.0, 1.0), attributes={"state": "ok"})]

gauge = meter.create_observable_gauge(
    name="example.gauge",
    callbacks=[observable_gauge_callback],
    description="Observable gauge metric",
    unit="1"
)

def observable_counter_callback(options):
    return [metrics.Observation(value=random.randint(100, 500), attributes={"type": "batch"})]

obs_counter = meter.create_observable_counter(
    name="example.observable_counter",
    callbacks=[observable_counter_callback],
    description="Observable counter",
    unit="1"
)

def observable_updown_callback(options):
    return [metrics.Observation(value=random.randint(-10, 10), attributes={"session": "live"})]

obs_updown = meter.create_observable_up_down_counter(
    name="example.observable_updowncounter",
    callbacks=[observable_updown_callback],
    description="Observable updowncounter",
    unit="connections"
)

if __name__ == "__main__":
    while True:
        request_counter.add(1, {"endpoint": "/api", "status": "200"})
        delta = random.choice([-1, 1])
        connections.add(delta, {"region": "eu-west"})
        response_times.record(random.gauss(120, 20), {"path": "/search"})

        time.sleep(2)
```

pront reviewed (approved):
All changes look good! Let's get a quick review from @vectordotdev/documentation as well.

aliciascott commented on src/sources/opentelemetry/integration_tests.rs:
```diff
@@ -182,6 +187,108 @@ async fn receive_trace() {
+                schema_url: "".to_string(),
+                metrics: vec![Metric {
+                    name: "some.random.metric".to_string(),
+                    description: "Some random metric we use for test".to_string(),
```
```suggestion
                    description: "Some random metric we use for test".to_string(),
```
Just want to be sure this is the intended description here ?

aliciascott replied:
I synced with @pront on this and understand this is not user facing content so I will dismiss this review, sorry for the confusion !

aliciascott reviewed (changes requested):
@cmcmacs there are several instances in the `description` string where I just want to be sure that you intended to use the literal value "some random metric to test" ?

aliciascott reviewed (approved):
Approved for docs

fbs commented:
good to see progress on otel metrics!

I only had a quick look at this, but it seems like this only supports cumulative metrics, right?

pront commented on src/sources/opentelemetry/tests.rs:
```diff
@@ -338,6 +353,997 @@ async fn receive_grpc_logs_legacy_namespace() {
+#[tokio::test]
+async fn receive_summary_metric() {
+    assert_source_compliance(&SOURCE_TAGS, async {
+        let grpc_addr = next_addr();
+        let http_addr = next_addr();
+
+        let source = OpentelemetryConfig {
+            grpc: GrpcConfig {
+                address: grpc_addr,
+                tls: Default::default(),
+            },
+            http: HttpConfig {
+                address: http_addr,
+                tls: Default::default(),
+                keepalive: Default::default(),
+                headers: Default::default(),
+            },
+            acknowledgements: Default::default(),
+            log_namespace: Default::default(),
+        };
+
+        let (sender, metrics_output, _) = new_source(EventStatus::Delivered, METRICS.to_string());
+        let server = source
+            .build(SourceContext::new_test(sender, None))
+            .await
+            .unwrap();
+        tokio::spawn(server);
+        test_util::wait_for_tcp(grpc_addr).await;
+
+        // send request via grpc client
+        let mut client = MetricsServiceClient::connect(format!("http://{}", grpc_addr))
```
First of all, thank you for providing all these tests!

Can de-deduplicate a bit? There's a lot of repeated boilerplate.

cmcmacs replied:
> First of all, thank you for providing all these tests!
> 
> Can de-deduplicate a bit? There's a lot of repeated boilerplate.

thanks, took server definition out of each test, apart from http headers, since they need to manipulate headers on the server side. and also time setting also now a function. Test funcs are now mostly object definitions of actual/expected.

cmcmacs commented:
> good to see progress on otel metrics!
> 
> I only had a quick look at this, but it seems like this only supports cumulative metrics, right?

@fbs @pront I took a deeper look if we can do both delta and cumulative, and added support for both. So now the strategy for mapping is:
if aggregation temporality supported by metric in OTel(sum, histogram and exponential histogram) and is set to Delta, then we map this metricKind as Incremental in Vector. Otherwise agg.temporality is not Delta and so metricKind in Vector would be Absolute. (Ref with info https://opentelemetry.io/docs/specs/otel/metrics/data-model/#temporality)

And sum is now mapped to gauge if is_monotonic is false, otherwise mapped to count. (following this diagram https://opentelemetry.io/docs/specs/otel/metrics/img/model-layers-stream.png)

Added to docs and tests.

pront commented:
> And sum is now mapped to gauge if is_monotonic is false, otherwise mapped to count. (following this diagram https://opentelemetry.io/docs/specs/otel/metrics/img/model-layers-stream.png)

This mapping seems reasonable based on https://opentelemetry.io/docs/specs/otel/metrics/data-model/#point-kinds.

pront commented:
@cmcmacs one last question before we can merge this. Did you have a chance to re-test with the latest version of this PR? Also, worth checking with both Absolute and Incremental metrics. Thanks again for this awesome contribution.

cmcmacs commented:
> @cmcmacs one last question before we can merge this. Did you have a chance to re-test with the latest version of this PR? Also, worth checking with both Absolute and Incremental metrics. Thanks again for this awesome contribution.

Tested with emiting metrics and it seems to be working fine, i tested both incremental and absolute, looks fine.

pront commented:
Hmm, macos tests timed out: https://github.com/vectordotdev/vector/actions/runs/14450435408/job/40521543264

cmcmacs commented:
I was testing it on Mac and it’s working fine, will look into it.

pront commented:
> I was testing it on Mac and it’s working fine, will look into it.

Giving this a second try in case it was a flake. If it fails again, you can look at the failure report for hints.
